evm/p2p/cancel_token.py

```diff
@@ -0,0 +1,90 @@
+import asyncio
+from typing import Any, Awaitable
+
+from evm.p2p.exceptions import OperationCancelled
+
+
+class CancelToken:
+
+    def __init__(self, name: str, next_token: 'CancelToken' = None) -> None:
+        self.name = name
+        self.next = next_token
+        self._triggered = asyncio.Event()
+
+    def chain(self, next_token: 'CancelToken') -> 'CancelToken':
+        """Return a new CancelToken chaining this and next_token.
+
+        The new CancelToken's triggered will return True if trigger() has been
+        called on either of the chained tokens, but calling trigger() on this token
+        has no effect on either of the chained tokens.
+        """
+        return CancelToken(":".join([self.name, next_token.name]), next_token)
+
+    def trigger(self) -> None:
+        self._triggered.set()
+
+    @property
+    def triggered_token(self) -> 'CancelToken':
+        if self._triggered.is_set():
+            return self
+        elif self.next is not None:
+            return self.next.triggered_token
+        else:
+            return None
+
+    @property
+    def triggered(self) -> bool:
+        if self.next is None:
+            return self._triggered.is_set()
+        return self._triggered.is_set() or self.next.triggered
+
+    async def wait(self) -> None:
+        if self.triggered_token is not None:
+            return
+
+        futures = [asyncio.ensure_future(self._triggered.wait())]
+        if self.next is not None:
+            futures.append(asyncio.ensure_future(self.next.wait()))
+
+        def cancel_pending(f):
+            for future in futures:
+                if not future.done():
+                    future.cancel()
+
+        fut = asyncio.ensure_future(self._wait_for_first(futures))
+        fut.add_done_callback(cancel_pending)
+        await fut
+
+    async def _wait_for_first(self, futures):
```

Comment:
Can this be a standalone utils function or does the `async` mean that it has to be part of the CancelToken class?

Reply:
the `async def` is what makes it a coroutine which is in-turn what allows you to call `await token._wait_for_first(...)`.

The `await` is what yields execution to other processes allow them to continue execution while this function would normally block.

@gsalgado can probably correct me if I'm wrong here.

Reply:
@pipermerriam's explanation is spot on, but this method can indeed be a standalone function, although I'd prefer to keep it in this module for now